frontends/p4/typeChecking/typeCheckExpr.cpp

```diff
@@ -2157,12 +2157,14 @@ const IR::Node *TypeInferenceBase::postorder(const IR::MethodCallExpression *exp
             const IR::Type *baseReturnType = returnType;
             if (const auto *sc = returnType->to<IR::Type_SpecializedCanonical>())
                 baseReturnType = sc->baseType;
-            const bool factoryOrStaticAssert =
-                baseReturnType->is<IR::Type_Extern>() || ef->method->name == "static_assert";
-            if (constArgs && factoryOrStaticAssert) {
+            const bool factoryOrStaticAssertOrPureAnnot =
+                baseReturnType->is<IR::Type_Extern>() || ef->method->name == "static_assert" ||
+                ef->method->getAnnotation(IR::Annotation::pureAnnotation);
+            if (constArgs && factoryOrStaticAssertOrPureAnnot) {
                 // factory extern function calls (those that return extern objects) with constant
                 // args are compile-time constants.
-                // The result of a static_assert call is also a compile-time constant.
+                // The result of a static_assert call is also a compile-time constant
+                // Pure functions marked with @pure annotation is a compiled-time constant
                 setCompileTimeConstant(expression);
                 setCompileTimeConstant(getOriginal<IR::Expression>());
             }
```

Comment:
I am no expert on `@pure`, but according to @ChrisDodd and @vlstill's previous comments (https://github.com/p4lang/p4c/pull/4726#issuecomment-2205808456, https://github.com/p4lang/p4c/pull/4726#issuecomment-2206037791, https://github.com/p4lang/p4c/pull/4726#issuecomment-2207204966), a pure extern function that has const args is not obligated to evaluate to a compile-time constant. Excluding such extern functions here was intentional in https://github.com/p4lang/p4c/pull/4726. @ChrisDodd Can you double-check that this makes sense?

Reply:
If you look at the example in this PR, would @consteval make more sense to you?

Reply:
Yes, some `consteval` keyword or annotation makes the most sense to me based on my understanding of the previous comments from @vlstill and @ChrisDodd. @vlstill @ChrisDodd What do you think?

Reply:
I'm of mixed mind about this.  `@pure` has a particular meaning (basically, the function depends only on its arguments and has no side effects), so any `@pure` function where all the arguments are constants is (logically) a constant.  But the term "compile-time constant" has a particular meaning in the spec, so it certainly seems plausible to have non-`@pure` functions that are also compile-time-constants if all the arguments are compile-time-constants.  In particular, extern functions that do have some sort of run-time side effect, but can otherwise be treated as compile time constants, and should NOT be dead-code eliminated if their result is not used (which is an allowed optimization for `@pure` functions).

For this specific case, are the functions in question ones that should not be subject to dead-code elimination if the value they return is unused?  If so, they should not be `@pure` and we need to come up with something else.

Reply:
Yes, dead-code eliminating the function if not used is fine.

Reply:
So I've been reading through the code, trying to figure out everywhere that this `isCompileTimeConstant` attribute of expressions stored in the typeMap is used and what it is used for.

For the most part, it is used for checking if things are "compile-time known values" as defined by the spec, and issuing errors for things that should be compile-time known and are not.  Not sure how the difference between "local compile-time known" and "compile-time known" applies here (important difference in the spec; not sure how it is reflected in the p4c code).  The spec doesn't really say if the result of an extern function can be compile-time known or not; IMO it makes sense to leave it up to the arch (since the arch defines what extern functions are and what they mean), so an annotation on the extern function declaration makes sense here.

However, it also appears to be used in cases like `sideEffects` where it is a short-cut for enabling optimizations -- a value that `isCompileTimeConstant` is true is assumed to not be alias with modifications or otherwise interact with things that have side effects, so can be reordered with respect to them.  The `@pure` annotation makes perfect sense for this case.

Not clear to me if we should think about breaking up these two cases.  We should definitely be more clear about what exactly is compile-time know vs local compile-time known vs known to be a constant for optimization purposes, but that is something that belongs in its own issue.   For the purpose of this change, I'm fine with overloading `@pure` to manage this, or with defining a new annotation if that is what people think is better.